index/index_monitor.go

```diff
@@ -0,0 +1,46 @@
+package index
+
+// helpers to atomically switch from using one index to another
+
+import (
+	"sync"
+)
+
+type IndexMonitor struct {
+	rwMutex sync.RWMutex
+	index *Index
+}
+
+func (im *IndexMonitor) AcquireRef() *Index {
+	im.rwMutex.RLock()
+	index := im.index
+	if index != nil {
+		index.refcount.Add(1)
+	}
+	im.rwMutex.RUnlock()
```

Comment:
`defer`? Not that it's likely to matter, but guaranteed cleanup mechanics >> theoretically non-guaranteed cleanup mechanics.